lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/MetricsRecordBuilder.cs

```diff
@@ -0,0 +1,186 @@
+        /// <returns>Self reference.</returns>
+        public IMetricsRecordBuilder Add(MetricsTag tag)
+        {
+            _tags.Add(tag);
+            return this;
+        }
+        
+        /// <summary>
+        /// Adds a metric to the record builder. This functions removes duplication in 
+        /// case metric has already been created by the caller.
+        /// </summary>
+        /// <param name="metric">Metric to add.</param>
+        /// <returns>Self reference.</returns>
+        public IMetricsRecordBuilder Add(IImmutableMetric metric)
+        {
+            _metrics.Add(metric);
+            return this;
+        }
+
+        /// <summary>
+        /// Sets a context for the record builder. Creates the corresponding tag.
+        /// </summary>
+        /// <param name="value">Value of the context.</param>
+        /// <returns>Self reference.</returns>
+        public IMetricsRecordBuilder SetContext(string value)
+        {
+            _contextValue = value;
+            return AddTag(MetricsSystemConstants.Context, value);
+        }
+
+        /// <summary>
+        /// Adds a counter to the record builder.
+        /// </summary>
+        /// <param name="info">Meta data info of the counter.</param>
+        /// <param name="value">Value of the counter.</param>
+        /// <returns>Self reference.</returns>
+        public IMetricsRecordBuilder AddCounter(IMetricsInfo info, long value)
+        {
+            _metrics.Add(new ImmutableCounter(info, value));
+            return this;
+        }
+
+        /// <summary>
+        /// Adds a long gauge to the record builder.
+        /// </summary>
+        /// <param name="info">Meta data info of the gauge.</param>
+        /// <param name="value">Value of the gauge.</param>
+        /// <returns>Self reference.</returns>
+        public IMetricsRecordBuilder AddGauge(IMetricsInfo info, long value)
+        {
+            _metrics.Add(new ImmutableLongGauge(info, value));
+            return this;
+        }
+
+        /// <summary>
+        /// Adds a double gauge to the record builder.
+        /// </summary>
+        /// <param name="info">Meta data info of the gauge.</param>
+        /// <param name="value">Value of the gauge.</param>
+        /// <returns>Self reference.</returns>
+        public IMetricsRecordBuilder AddGauge(IMetricsInfo info, double value)
+        {
+            _metrics.Add(new ImmutableDoubleGauge(info, value));
+            return this;
+        }
+
+        /// <summary>
+        /// Returns the parent collector.
+        /// </summary>
+        public IMetricsCollector ParentCollector()
+        {
+            return _parentCollector;
+        }
+
+        /// <summary>
+        /// Ends the record building and returns the parent collector.
+        /// </summary>
+        /// <returns>The parent collector.</returns>
+        public IMetricsCollector EndRecord()
+        {
+            return _parentCollector;
```

Comment:
How does this implementation ensure that calling this method ends record building? I don't see anything preventing user from adding more things to the record.